src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -655,6 +655,12 @@ internal void InitializeForReflectionSerializer()
         private JsonTypeInfo? GetTypeInfoInternal(Type type)
         {
             IJsonTypeInfoResolver? resolver = _effectiveJsonTypeInfoResolver ?? _typeInfoResolver;
+
+            if (resolver == null)
+            {
+                ThrowHelper.ThrowInvalidOperationException_JsonTypeInfoUsedButTypeInfoResolverNotSet();
```

Comment:
That doesn't look like the appropriate location to throw this error message. I think this method can be invoked from other contexts as well, e.g. when doing `options.GetConverter().Write(...)`

Reply:
I just noticed that and it broke some tests

Reply:
I think it should be moved to the root-level serialization methods that accept JsonTypeInfo